src/Functions/if.cpp

```diff
@@ -656,30 +656,83 @@ class FunctionIf : public FunctionIfBase</*null_is_false=*/false>
     }
 
-    bool executeForNullableCondition(Block & block, const ColumnNumbers & arguments, size_t result, size_t /*input_rows_count*/)
+    bool executeForConstAndNullableCondition(Block & block, const ColumnNumbers & arguments, size_t result, size_t /*input_rows_count*/)
     {
         const ColumnWithTypeAndName & arg_cond = block.getByPosition(arguments[0]);
         bool cond_is_null = arg_cond.column->onlyNull();
 
-        if (cond_is_null)
+        ColumnPtr not_const_condition = arg_cond.column;
+        bool cond_is_const = false;
+        bool cond_is_true = false;
+        bool cond_is_false = false;
+        if (const auto * const_arg = checkAndGetColumn<ColumnConst>(*arg_cond.column))
         {
-            block.getByPosition(result).column = std::move(block.getByPosition(arguments[2]).column);
-            return true;
+            cond_is_const = true;
+            not_const_condition = const_arg->getDataColumnPtr();
+            ColumnPtr data_column = const_arg->getDataColumnPtr();
+            if (const auto * const_nullable_arg = checkAndGetColumn<ColumnNullable>(*data_column))
+            {
+                data_column = const_nullable_arg->getNestedColumnPtr();
+                if (!data_column->empty())
+                    cond_is_null = const_nullable_arg->getNullMapData()[0];
+            }
+
+            if (!data_column->empty())
+            {
+                cond_is_true = !cond_is_null && checkAndGetColumn<ColumnUInt8>(*data_column)->getBool(0);
+                cond_is_false = !cond_is_null && !cond_is_true;
+            }
+        }
+
+        const auto & column1 = block.getByPosition(arguments[1]);
+        const auto & column2 = block.getByPosition(arguments[2]);
+        auto & result_column = block.getByPosition(result);
+
+        if (cond_is_true)
+        {
+            if (result_column.type->equals(*column1.type))
+            {
+                result_column.column = std::move(column1.column);
+                return true;
+            }
+        }
+        else if (cond_is_false || cond_is_null)
+        {
+            if (result_column.type->equals(*column2.type))
+            {
+                result_column.column = std::move(column2.column);
+                return true;
+            }
         }
 
-        if (const auto * nullable = checkAndGetColumn<ColumnNullable>(*arg_cond.column))
+        if (const auto * nullable = checkAndGetColumn<ColumnNullable>(*not_const_condition))
         {
+            ColumnPtr new_cond_column = nullable->getNestedColumnPtr();
+            size_t column_size = arg_cond.column->size();
+
+            if (cond_is_null) /// Nullable(Nothing)
+                new_cond_column = ColumnConst::create(ColumnUInt8::create(1, 0), column_size);
+            else if (checkAndGetColumn<ColumnUInt8>(*new_cond_column))
+            {
+                new_cond_column = nullable->getNestedColumnCopy(); /// zero NULLs in ColumnUInt8
```

Comment:
If we know that column is `Nullable(UInt8)`, we can write simple loop which calculates logical `and` even right here.
Is it always true, btw?

Reply:
> Is it always true, btw?

It seems to be so.

Reply:
The source column is const here. It's not clear if we could change nested column inplace. I've created inplace `ColumnVector::applyZeroMap()` instead of `getNestedColumnCopy()`. We could make optimisation for Uint8 with logical `and` later. It think there would be not much profit of optimisation here cause we cannot avoid column copy.

Reply:
We *almost* can avoid copying.
Right now we can call `IColumn::mutate()` for filter, and if it has `ref_count = 1`, it won't be copied.
The only problem that we have not information if we need this column later. And functions can't change arguments in block in current interface.

Reply:
Yup, there's posibility to change but it's dangerous cause contradicts the function interface. There're no such guaranties here.